packages/typespec-python/tasks.py

```diff
@@ -114,92 +113,149 @@
-    name = _package_name_folder(spec)
+def _default_package_name(spec: Path, specification_dirs: List[Path]) -> str:
+    return _package_name_folder(spec, specification_dirs).replace("/", "-")
+
+
+def _get_emitter_option(
+    spec: Path, specification_dirs: List[Path]
+) -> List[Dict[str, str]]:
+    name = _package_name_folder(spec, specification_dirs)
     result = EMITTER_OPTIONS.get(name, [])
     if isinstance(result, dict):
         return [result]
     return result
 
 
-def _add_options(spec: Path, debug=False) -> List[str]:
+def _add_options(
+    spec: Path,
+    specification_dirs: List[Path],
+    generated_foder: Path,
+    special_flags: Dict[str, Any],
+    debug=False,
+) -> List[str]:
     # if debug:
     #   options["debug"] = "true"
     result = []
-    for config in _get_emitter_option(spec):
+    for config in _get_emitter_option(spec, specification_dirs):
         config_copy = copy.copy(config)
-        config_copy["emitter-output-dir"] = f"{PLUGIN_DIR}/test/generated/{config['package-name']}"
+        config_copy[
+            "emitter-output-dir"
+        ] = f"{generated_foder}/{config['package-name']}"
         result.append(config_copy)
     if not result:
-        result.append({"emitter-output-dir": f"{PLUGIN_DIR}/test/generated/{_default_package_name(spec)}"})
-    return [" --option ".join(
-        [f"@azure-tools/typespec-python.{k}={v} " for k, v in options.items()]
-    ) for options in result]
+        result.append(
+            {
+                "emitter-output-dir": f"{generated_foder}/{_default_package_name(spec, specification_dirs)}"
+            }
+        )
+    emitter_configs = []
+    for options in result:
+        emitter_option = ""
+        for item in [options, special_flags]:
+            for k, v in item.items():
+                emitter_option += f" --option @azure-tools/typespec-python.{k}={v}"
+        emitter_configs.append(emitter_option)
+    return emitter_configs
 
 
 def _entry_file_name(path: Path) -> Path:
     if path.is_file():
         return path
-    return (path / "client.tsp") if (path / "client.tsp").exists() else (path / "main.tsp")
+    return (
+        (path / "client.tsp") if (path / "client.tsp").exists() else (path / "main.tsp")
+    )
 
-def all_specification_folders() -> List[Path]:
-    return [s for item in ALL_SPECIFICATION_DIRS for s in item.glob("**/*") if s.is_dir()]
 
-@task
-def regenerate(c, name=None, debug=False):
+def all_specification_folders(specification_dirs: List[Path]) -> List[Path]:
+    return [s for item in specification_dirs for s in item.glob("**/*") if s.is_dir()]
+
+
+def _regenerate(c, test_config, name=None, debug=False):
+    generated_sub_folder = test_config["generated_sub_folder"]
+    local_specification_folder = Path(f"test/{generated_sub_folder}/specification")
+    specification_dirs = (
+        [CADL_RANCH_DIR, local_specification_folder]
+        if local_specification_folder.exists()
```

Comment:
our logic is getting too gross with generating mgmt sphere. I think we should treat mgmt sphere like we treat storage from swagger: after we merge a PR we should regenerate with mgmt sphere tsp, so we make sure we don't break it on release. Then we can get rid of the whole local specification folder. I'm good to do these in tangent with each other.

I also think we should specifically have two functions: `regenerate_azure` and `regenerate_unbranded`, we shouldn't abstract this away. Then `regenerate` will just have two lines: one for calling each one. The reasons are because we expect these two to be the only ones, and this way, we can also call `inv regenerate-azure` or `inv regenerate-unbranded` from the command line, which provides good support

Reply:
(1) Agree that we shall generate from swagger repo directly. Let me do it later.
(2) Will create 2 functions`regenerate_azure` and `regenerate_unbranded`